client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx

```diff
@@ -11,9 +16,31 @@ type StepRouteProps = {
 	renderStep: ( step: StepperStep ) => JSX.Element | null;
 };
 
+//TODO: Check we can move RenderStep function to here and remove the renderStep prop
 const StepRoute = ( { step, flow, showWooLogo, renderStep }: StepRouteProps ) => {
+	const userIsLoggedIn = useSelector( isUserLoggedIn );
+	const loginUrl = useLoginUrlForFlow( { flow } );
 	const stepContent = renderStep( step );
+	const shouldRedirectToLogin = step.requiresLoggedInUser && ! userIsLoggedIn;
+	const shouldSkipRender = shouldRedirectToLogin || ! stepContent;
 
+	useStepRouteTracking( {
+		flowName: flow.name,
+		stepSlug: step.slug,
+		skipTracking: shouldSkipRender,
+	} );
+
+	useEffect( () => {
+		if ( shouldRedirectToLogin ) {
+			window.location.assign( loginUrl );
+		}
+	}, [ loginUrl, shouldRedirectToLogin ] );
+
+	if ( shouldSkipRender ) {
+		return null;
+	}
+
+	//TODO: Extract the layout logics to another component;
```

Comment:
I am not sure what you mean by this comment. Could you be more explicit about which layout logic should be moved/refactored?

Reply:
Comment removed [e576a62](https://github.com/Automattic/wp-calypso/pull/91241/commits/e576a62fabfcf7292ce645b26f784a6a4c19c3c9)